src/wallet/verify.rs

```diff
@@ -29,45 +29,57 @@ use crate::error::Error;
 /// Depending on the [capabilities](crate::blockchain::Blockchain::get_capabilities) of the
 /// [`Blockchain`] backend, the method could fail when called with old "historical" transactions or
 /// with unconfirmed transactions that have been evicted from the backend's memory.
-pub fn verify_tx<D: Database, B: Blockchain>(
-    tx: &Transaction,
-    database: &D,
-    blockchain: &B,
-) -> Result<(), VerifyError> {
-    log::debug!("Verifying {}", tx.txid());
-
-    let serialized_tx = serialize(tx);
-    let mut tx_cache = HashMap::<_, Transaction>::new();
-
-    for (index, input) in tx.input.iter().enumerate() {
-        let prev_tx = if let Some(prev_tx) = tx_cache.get(&input.previous_output.txid) {
-            prev_tx.clone()
-        } else if let Some(prev_tx) = database.get_raw_tx(&input.previous_output.txid)? {
-            prev_tx
-        } else if let Some(prev_tx) = blockchain.get_tx(&input.previous_output.txid)? {
-            prev_tx
-        } else {
-            return Err(VerifyError::MissingInputTx(input.previous_output.txid));
-        };
-
-        let spent_output = prev_tx
-            .output
-            .get(input.previous_output.vout as usize)
-            .ok_or(VerifyError::InvalidInput(input.previous_output))?;
-
-        bitcoinconsensus::verify(
-            &spent_output.script_pubkey.to_bytes(),
-            spent_output.value,
-            &serialized_tx,
-            index,
-        )?;
-
-        // Since we have a local cache we might as well cache stuff from the db, as it will very
-        // likely decrease latency compared to reading from disk or performing an SQL query.
-        tx_cache.insert(prev_tx.txid(), prev_tx);
-    }
 
-    Ok(())
+impl<B, D> Wallet<B, D>
+where
+    B: Blockchain,
+    D: BatchDatabase,
+{
+    /// Performs script verification for a given transaction.
+    /// This uses bitcoinconsensus.
+    /// It can be used as an wallet tool to verify custom transactions by bitcoin's consensus rules.
+    ///
+    /// Note: All blockchains won't be able to verify all transactions.
+    ///
+    /// `electrum` and `esplora` support (confed/unconfed) *any* transaction verification,
+    /// `rpc` supports (confed/unconfed) *in-wallet* transaction verification
+    /// `compact_filters` supports (unconfed) *any* transaction verification.
+    pub fn verify_tx(&self, tx: &Transaction) -> Result<(), VerifyError> {
```

Comment:
After thinking about this further I really think we should remove this API for now. 

If a user wants to verify a transaction it's unlikely they want to verify a transaction that's in the database. If it's in the database they can just check if it's in the database. If it is then it's verified (assuming you compiled with verification on).

If it's not in the database then it's really a blockchain specific thing in which case it should be a trait that is implemented on blockchains.

Reply:
I think I get what you are saying. 

Even if its an inherent `Blockchain` thing, we don't want to force user with its impl in their custom backends. So one way to do that will be to create a default implementation for any `Blockchain`. This can then be feature gated and blanket implemented. 

And the end result would look like `wallet.blockchain.verify()`, instead of current `wallet.verify()`. I am not sure if thats a big enough API win.

My envision of using the verify API is mostly for non wallet transactions, because its a reasonable assumption that you trust your own DB. And they can enforce it at sync level with `--verify` feature flag for that purpose.

But there should be an easy way to verify non wallet txs. (Even random message verification at future). So the library needs the API somewhere.

So I think for this PR I will just let it be here. And we can make the move in future PR if we want.

Reply:
@LLFourn  Do you feel satisfied with this reasoning?

Reply:
@rajarshimaitra I'm not sure.

I've thought about this a lot and I think that transaction verification is a very context specific thing and I'm not sure this function fits any of those contexts. Do you have one in mind?

In what situations do you need to verify non-wallet transactions where you don't care about whether the transaction is actually valid but just f the inputs exists it passes the consensus rules. Don't you also want to check if the inputs has have been spent or otherwise exist or if the locktime on the tx has passed? Also note this function fails with backends without a transaction index. How are you meant to deal with this failure in the context where it is useful (which again I can't conceive of).

I can understand wanting to check a tx is valid without broadcasting it. Bitcoin core rpc [has a call](https://developer.bitcoin.org/reference/rpc/testmempoolaccept.html) for this. With esplora it would be a matter of checking whether the inputs have been spent and using libconsensus. So I see this as a blockchain specific check which this function doesn't implement properly. It may fail depending on the backend which is odd because the backend it fails on (rpc) is the only one that has a call to do this thing properly (testmempoolaccept).

I don't think it's worth holding up this PR on this but if we are unsure about this function then it makes sense to me to leave it where it is. Why break the API only to break it again later on?

Reply:
Hmm, I guess I see what you mean. My initial use case was, if I build a transaction with a custom script P2SH and I create my custom witness to spend it, a easy way to check if the the witness works would be useful. Something like `testmempoolaccept` of core RPC.

I don't have any strong opinion on where this should be placed, and no solution for the case of non-txindex blockchains. 

Currently its not breaking the wallet API. Using the lib with `--verify` flag will just open this extra function call on wallet, no other existing calls changes.

What do you suggest is the way forward with this PR?

Reply:
> Hmm, I guess I see what you mean. My initial use case was, if I build a transaction with a custom script P2SH and I create my custom witness to spend it, a easy way to check if the the witness works would be useful. Something like `testmempoolaccept` of core RPC.

Wouldn't using libconsensus for the P2SH output be preferable in this case? testmempool would only be useful if you want to check the outputs actually exist at the same time and the network time allows for it etc. Another option is to use [`Interpreter`](https://docs.rs/miniscript/6.0.1/miniscript/interpreter/struct.Interpreter.html#method.from_txdata) from rust-miniscript.

> Currently its not breaking the wallet API. Using the lib with `--verify` flag will just open this extra function call on wallet, no other existing calls changes.

It's breaking the API because it removes the existing function.

> What do you suggest is the way forward with this PR?

I'd suggest the easiest way forward is to remove the changes related to the verify function from this PR and then make an issue or follow up PR where we can discuss whether this API is even a good idea or not. iirc @afilini introduced this function so his input is probably more useful than mine here.